controller/api/destination/watcher/cluster_watcher.go

```diff
@@ -0,0 +1,320 @@
+			if !found {
+				cs.log.Tracef("Skipping Add event for 'Secret' object %s/%s: missing \"%s\" label", secret.Namespace, secret.Name, clusterNameLabel)
+				return
+			}
+
+			if err := cs.addWatcher(clusterName, secret); err != nil {
+				cs.log.Errorf("Error adding watcher for cluster %s: %v", clusterName, err)
+			}
+
+		},
+		DeleteFunc: func(obj interface{}) {
+			secret, ok := obj.(*v1.Secret)
+			if !ok {
+				// If the Secret was deleted when the watch was disconnected
+				// (for whatever reason) and the event was missed, the object is
+				// added with 'DeletedFinalStateUnknown'. Its state may be
+				// stale, so it should be cleaned-up.
+				tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+				if !ok {
+					cs.log.Debugf("unable to get object from DeletedFinalStateUnknown %#v", obj)
+					return
+				}
+				// If the zombie object is a `Secret` we can delete it.
+				secret, ok = tombstone.Obj.(*v1.Secret)
+				if !ok {
+					cs.log.Debugf("DeletedFinalStateUnknown contained object that is not a Secret %#v", obj)
+					return
+				}
+			}
+
+			clusterName, found := secret.GetLabels()[clusterNameLabel]
+			if !found {
+				cs.log.Tracef("Skipping Delete event for 'Secret' object %s/%s: missing \"%s\" label", secret.Namespace, secret.Name, clusterNameLabel)
+				return
+			}
+
+			cs.removeWatcher(clusterName)
+
+		},
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	return cs, nil
+}
+
+// Get safely retrieves a watcher from the cache given a cluster name. It
+// returns the watcher, cluster configuration, if the entry exists in the cache.
+func (cs *ClusterStore) Get(clusterName string) (*EndpointsWatcher, string, string, bool) {
+	cs.RLock()
+	defer cs.RUnlock()
+	s, found := cs.store[clusterName]
+	return s.watcher, s.trustDomain, s.clusterDomain, found
+}
+
+// GetWatcher is a convenience method that given a cluster name only returns the
+// associated EndpointsWatcher if it exists in the cache.
+func (cs *ClusterStore) GetWatcher(clusterName string) (*EndpointsWatcher, bool) {
+	cs.RLock()
+	defer cs.RUnlock()
+	s, found := cs.store[clusterName]
+	return s.watcher, found
+}
+
+// GetLocalWatcher is a convenience method that retrieves the watcher associated
+// with the local cluster. Its existence is assumed.
+func (cs *ClusterStore) GetLocalWatcher() *EndpointsWatcher {
+	cs.RLock()
+	defer cs.RUnlock()
+	return cs.store[LocalClusterKey].watcher
+}
+
+// GetClusterConfig is a convenience method that given a cluster name retrieves
+// its associated configuration strings if the entry exists in the cache.
+func (cs *ClusterStore) GetClusterConfig(clusterName string) (string, string, bool) {
+	cs.RLock()
+	defer cs.RUnlock()
+	s, found := cs.store[clusterName]
+	return s.trustDomain, s.clusterDomain, found
+}
+
+// AddLocalWatcher adds a watcher to the cache using the local cluster key.
+func (cs *ClusterStore) AddLocalWatcher(stopCh chan<- struct{}, watcher *EndpointsWatcher, trustDomain, clusterDomain string) {
```

Comment:
Is there a reason we need to manage the local EndpointsWatcher through this store?  the local watcher doesn't go through any of the same secrets-based lifecycle, so I think it might be easier to keep it entirely separate.

Reply:
nope, no reason other than it seemed to fit. Keeping it entirely separate sounds good to me and might be easier to wire up.